cmd/kubeadm/app/util/etcd/etcd_test.go

```diff
@@ -106,3 +115,285 @@ func TestGetClientURLByIP(t *testing.T) {
 		}
 	}
 }
+
+func TestGetEtcdEndpointsWithBackoff(t *testing.T) {
+	var tests = []struct {
+		name              string
+		pods              []testresources.FakeStaticPod
+		configMap         *testresources.FakeConfigMap
+		expectedEndpoints []string
+		expectedErr       bool
+	}{
+		{
+			name:              "no pod annotations; no ClusterStatus",
+			expectedEndpoints: []string{},
+		},
+		{
+			name: "ipv4 endpoint in pod annotation; no ClusterStatus; port is preserved",
+			pods: []testresources.FakeStaticPod{
+				{
+					Component: constants.Etcd,
+					Annotations: map[string]string{
+						constants.EtcdAdvertiseClientUrlsAnnotationKey: "https://1.2.3.4:1234",
+					},
+				},
+			},
+			expectedEndpoints: []string{"https://1.2.3.4:1234"},
+		},
+		{
+			name:              "no pod annotations; ClusterStatus with valid ipv4 endpoint; port is inferred",
+			configMap:         testresources.ClusterStatusWithAPIEndpoint("cp-0", kubeadmapi.APIEndpoint{AdvertiseAddress: "1.2.3.4", BindPort: 1234}),
+			expectedEndpoints: []string{"https://1.2.3.4:2379"},
+		},
+	}
+	for _, rt := range tests {
+		t.Run(rt.name, func(t *testing.T) {
+			client := clientsetfake.NewSimpleClientset()
+			for _, pod := range rt.pods {
+				if err := pod.Create(client); err != nil {
+					t.Errorf("error setting up test creating pod for node %q", pod.NodeName)
+				}
+			}
+			if rt.configMap != nil {
+				if err := rt.configMap.Create(client); err != nil {
+					t.Error("could not create ConfigMap")
+				}
+			}
+			endpoints, err := getEtcdEndpointsWithBackoff(client, wait.Backoff{Duration: 0, Jitter: 0, Steps: 1})
+			if err != nil && !rt.expectedErr {
+				t.Errorf("got error %q; was expecting no errors", err)
+				return
+			} else if err == nil && rt.expectedErr {
+				t.Error("got no error; was expecting an error")
+				return
+			} else if err != nil && rt.expectedErr {
+				return
+			}
+
+			if !reflect.DeepEqual(endpoints, rt.expectedEndpoints) {
+				t.Errorf("expected etcd endpoints: %v; got: %v", rt.expectedEndpoints, endpoints)
+			}
+		})
+	}
+}
+
+func TestGetRawEtcdEndpointsFromPodAnnotation(t *testing.T) {
+	var tests = []struct {
+		name              string
+		backoff           wait.Backoff
+		pods              []testresources.FakeStaticPod
+		clientSetup       func(*clientsetfake.Clientset)
+		expectedEndpoints []string
+		expectedErr       bool
+	}{
+		{
+			name:    "exactly one pod with annotation",
+			backoff: wait.Backoff{Duration: 0, Jitter: 0, Steps: 1},
```

Comment:
nit, same as above (do not test ExponentialBackoff)

Reply:
Same, but with [this logic](https://github.com/kubernetes/kubernetes/pull/87656/files#diff-5e4c5bba67c635568de70e0c424882c6R131-R140).